File: Dairyncia-dotnet/Controllers/AdminController.cs
using Dairyncia.Data;
using Dairyncia.DTOs;
using Dairyncia.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace Dairyncia.Controllers
{
    [ApiController]
    [Route("api/admin")]
    //[Authorize(Roles = "Admin")]
    public class AdminController : ControllerBase
    {
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly RoleManager<IdentityRole> _roleManager;
        private readonly AppDbContext _context;

        public AdminController(
            UserManager<ApplicationUser> userManager,
            RoleManager<IdentityRole> roleManager,
            AppDbContext context)
        {
            _userManager = userManager;
            _roleManager = roleManager;
            _context = context;
        }

        // ================= ASSIGN ROLE =================
        [HttpPost("assign-role")]
        public async Task<IActionResult> AssignRole([FromBody] AssignRoleDto dto)
        {
            // 1️⃣ Find user
            var user = await _userManager.FindByEmailAsync(dto.Email);
            if (user == null)
                return NotFound("User not found");

            // 2️⃣ Validate role
            if (!await _roleManager.RoleExistsAsync(dto.Role))
                return BadRequest("Role does not exist");

            // 3️⃣ Remove existing roles (single-role system)
            var currentRoles = await _userManager.GetRolesAsync(user);
            if (currentRoles.Any())
                await _userManager.RemoveFromRolesAsync(user, currentRoles);

            // 4️⃣ Assign new role
            var roleResult = await _userManager.AddToRoleAsync(user, dto.Role);
            if (!roleResult.Succeeded)
                return BadRequest(roleResult.Errors);

            // 5️⃣ DOMAIN LOGIC (IMPORTANT)
            if (dto.Role.ToUpper() == "FARMER")
            {
                var farmerExists = await _context.Farmers
                    .AnyAsync(f => f.UserId == user.Id);

                if (!farmerExists)
                {
                    var farmer = new Farmer
                    {
                        UserId = user.Id,
                        AddressId = null // Address can be added later
                    };

                    _context.Farmers.Add(farmer);
                    await _context.SaveChangesAsync();
                }
            }

            // (Optional) If role changed from Farmer → remove Farmer row
            if (currentRoles.Contains("Farmer") && dto.Role.ToUpper() != "FARMER")
            {
                var farmer = await _context.Farmers
                    .FirstOrDefaultAsync(f => f.UserId == user.Id);

                if (farmer != null)
                {
                    _context.Farmers.Remove(farmer);
                    await _context.SaveChangesAsync();
                }
            }

            return Ok(new
            {
                message = $"Role '{dto.Role}' assigned successfully",
                user.Email
            });
        }

        // ================= PENDING USERS =================
        [HttpGet("pending-users")]
        public async Task<IActionResult> GetPendingUsers()
        {
            var pendingUsers = await (
                from user in _context.Users
                join userRole in _context.UserRoles
                    on user.Id equals userRole.UserId into ur
                from userRole in ur.DefaultIfEmpty()
                where userRole == null
                select new
                {
                    user.Id,
                    user.Email,
                    user.FullName,
                    user.PhoneNumber
                }
            ).ToListAsync();

            var result = pendingUsers.Select((u, index) => new PendingUserDto
            {
                SrNo = index + 1,
                UserId = u.Id,
                Email = u.Email,
                FullName = u.FullName,
                Phone = u.PhoneNumber
            });

            return Ok(result);
        }



        // list of farmers
        [HttpGet("farmers")]
        public async Task<IActionResult> GetAllFarmers()
        {
            var farmers = await _context.Farmers
            .Include(f => f.User)
            .Include(f => f.Address)
            .OrderByDescending(f => f.CreatedAt)
            .Select(f => new FarmerListDTO
            {
                FarmerId = f.Id,
                Email = f.User.Email,
                FullName = f.User.FullName,
                CreatedAt = f.CreatedAt
            })
            .ToListAsync();
            return Ok(farmers);
        }

        // famer details by id
        [HttpGet("farmers/{id}")]
        public async Task<IActionResult> GetFarmerDetails(int id)
        {
            var farmer = await _context.Farmers
                .Include(f => f.User)
                .Include(f => f.Address)
                .FirstOrDefaultAsync(f => f.Id == id);

            if (farmer == null)
                return NotFound("Farmer not found");

            var bank = await _context.BankDetails
                .FirstOrDefaultAsync(b => b.UserId == farmer.UserId);

            return Ok(new FarmerDetailsDto
            {
                FarmerId = farmer.Id,
                UserId = farmer.UserId,
                FullName = farmer.User.FullName,
                Email = farmer.User.Email,
                Phone = farmer.User.PhoneNumber,
                Address = farmer.Address == null ? null : new AddressDto
                {
                    AddressId = farmer.Address.Id,
                    Village = farmer.Address.Village,
                    City = farmer.Address.City,
                    State = farmer.Address.State,
                    Pincode = farmer.Address.Pincode
                },
                BankDetails = bank == null ? null : new BankDetailsDto
                {
                    BankDetailsId = bank.Id,
                    BankName = bank.BankName,
                    AccountNumber = bank.AccountNumber,
                    AccountHolderName = bank.AccountHolderName,
                    IFSC = bank.IFSC
                }
            });
        }

        // add or update address
        [HttpPost("farmers/{id}/address")]
        public async Task<IActionResult> AddOrUpdateAddress(int id, AddressDto dto)
        {
            var farmer = await _context.Farmers
                .Include(f => f.Address)
                .FirstOrDefaultAsync(f => f.Id == id);

            if (farmer == null)
                return NotFound("Farmer not found");

            if (farmer.Address == null)
            {
                var address = new Address
                {
                    Village = dto.Village,
                    City = dto.City,
                    State = dto.State,
                    Pincode = dto.Pincode
                };

                _context.Addresses.Add(address);
                await _context.SaveChangesAsync();

                farmer.AddressId = address.Id;
            }
            else
            {
                farmer.Address.Village = dto.Village;
                farmer.Address.City = dto.City;
                farmer.Address.State = dto.State;
                farmer.Address.Pincode = dto.Pincode;
            }

            await _context.SaveChangesAsync();
            return Ok("Address saved successfully");
        }

        //
        [HttpPost("farmers/{id}/bank")]
        public async Task<IActionResult> AddOrUpdateBank(int id, BankDetailsDto dto)
        {
            var farmer = await _context.Farmers.FindAsync(id);
            if (farmer == null)
                return NotFound("Farmer not found");

            var bank = await _context.BankDetails
                .FirstOrDefaultAsync(b => b.UserId == farmer.UserId);

            if (bank == null)
            {
                bank = new BankDetails
                {
                    UserId = farmer.UserId,
                    BankName = dto.BankName,
                    AccountNumber = dto.AccountNumber,
                    AccountHolderName = dto.AccountHolderName,
                    IFSC = dto.IFSC
                };
                _context.BankDetails.Add(bank);
            }
            else
            {
                bank.BankName = dto.BankName;
                bank.AccountNumber = dto.AccountNumber;
                bank.AccountHolderName = dto.AccountHolderName;
                bank.IFSC = dto.IFSC;
            }

            await _context.SaveChangesAsync();
            return Ok("Bank details saved successfully");
        }

        // delete farmer
        [HttpDelete("farmers/{id}")]
        public async Task<IActionResult> DeleteFarmer(int id)
        {
            var farmer = await _context.Farmers.FindAsync(id);
            if (farmer == null)
                return NotFound();

            var hasMilk = await _context.MilkCollections
                .AnyAsync(m => m.FarmerId == id);

            if (hasMilk)
                return BadRequest("Farmer cannot be deleted. Milk records exist.");

            _context.Farmers.Remove(farmer);
            await _context.SaveChangesAsync();

            return Ok("Farmer deleted");
        }

        // edit farmer profile
        [HttpPut("farmers/{id}")]
        public async Task<IActionResult> UpdateFarmer(int id, UpdateFarmerDto dto)
        {
            var farmer = await _context.Farmers
                .Include(f => f.User)
                .FirstOrDefaultAsync(f => f.Id == id);

            if (farmer == null)
                return NotFound("Farmer not found");

            farmer.User.FullName = dto.FullName;
            farmer.User.PhoneNumber = dto.PhoneNumber;

            await _context.SaveChangesAsync();
            return Ok("Farmer profile updated successfully");
        }






        // list of managers
        [HttpGet("managers")]
        public async Task<IActionResult> GetAllManagers()
        {
            var managers = await _userManager.GetUsersInRoleAsync("Manager");

            var result = managers
                .Select(u => new ManagerListDto
                {
                    ManagerId = u.Id,
                    Email = u.Email,
                    FullName = u.FullName,
                    Phone = u.PhoneNumber
                })
                .ToList();

            return Ok(result);
        }

        // get manager details by id
        [HttpGet("managers/{id}")]
        public async Task<IActionResult> GetManagerById(string id)
        {
            var manager = await _userManager.FindByIdAsync(id);
            if (manager == null)
                return NotFound("Manager not found");

            return Ok(new
            {
                manager.Id,
                manager.Email,
                manager.FullName,
                manager.PhoneNumber
            });
        }

        // update manager details
        [HttpPut("managers/{id}")]
        public async Task<IActionResult> UpdateManager(string id, UpdateManagerDto dto)
        {
            var manager = await _userManager.FindByIdAsync(id);
            if (manager == null)
                return NotFound("Manager not found");

            manager.FullName = dto.FullName;
            manager.PhoneNumber = dto.PhoneNumber;

            var result = await _userManager.UpdateAsync(manager);
            if (!result.Succeeded)
                return BadRequest(result.Errors);

            return Ok("Manager updated successfully");
        }

        // delete manager 
        [HttpDelete("managers/{id}")]
        public async Task<IActionResult> DeleteManager(string id)
        {
            var manager = await _userManager.FindByIdAsync(id);
            if (manager == null)
                return NotFound("Manager not found");

            // Check references
            var isUsed = await _context.MilkCollections
                .AnyAsync(m => m.ManagerId == id);

            if (isUsed)
                return BadRequest("Manager cannot be deleted. Records exist.");

            // Remove roles
            var roles = await _userManager.GetRolesAsync(manager);
            if (roles.Any())
                await _userManager.RemoveFromRolesAsync(manager, roles);

            var result = await _userManager.DeleteAsync(manager);
            if (!result.Succeeded)
                return BadRequest(result.Errors);

            return Ok("Manager deleted successfully");
        }

        // list of milk details
        [HttpGet]
        public async Task<IActionResult> GetAll()
        {
            var list = await _context.MilkCollections
                .Include(m => m.Farmer).ThenInclude(f => f.User)
                .Include(m => m.Manager)
                .OrderByDescending(m => m.CreatedAt)
                .Select(m => new MilkCollectionListDto
                {
                    Id = m.Id,
                    FarmerName = m.Farmer.User.FullName,
                    ManagerName = m.Manager.FullName,
                    MilkType = m.MilkType,
                    MilkShift = m.MilkShift,
                    Quantity = m.Quantity,
                    TotalAmount = m.TotalAmount,
                    PaymentStatus = m.PaymentStatus,
                    CreatedAt = m.CreatedAt
                })
                .ToListAsync();

            return Ok(list);
        }

        // update milk details
        [HttpPut("{id}")]
        public async Task<IActionResult> Update(int id, UpdateMilkCollectionDto dto)
        {
            var record = await _context.MilkCollections.FindAsync(id);
            if (record == null)
                return NotFound();

            record.Quantity = dto.Quantity;
            record.FatPercentage = dto.FatPercentage;
            record.SNF = dto.SNF;
            record.RatePerLiter = dto.RatePerLiter;
            record.TotalAmount = dto.Quantity * dto.RatePerLiter;
            record.PaymentStatus = dto.PaymentStatus;

            await _context.SaveChangesAsync();
            return Ok("Milk collection updated");
        }

        // delete milk record
        [HttpDelete("{id}")]
        public async Task<IActionResult> Delete(int id)
        {
            var record = await _context.MilkCollections.FindAsync(id);
            if (record == null)
                return NotFound();

            _context.MilkCollections.Remove(record);
            await _context.SaveChangesAsync();

            return Ok("Milk collection deleted");
        }


    }
}

File: Dairyncia-Spring/src/main/java/com/dairyncia/controllers/AdminController.java
package com.dairyncia.controllers;

import java.util.ArrayList;
import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.dairyncia.dto.AddressDTO;
import com.dairyncia.dto.ApiResponse;
import com.dairyncia.dto.AssignRoleDto;
import com.dairyncia.dto.BankDetailsDTO;
import com.dairyncia.dto.FarmerProfileUpdateDTO;
import com.dairyncia.dto.FarmerResponseDTO;
import com.dairyncia.dto.ManagerResponseDTO;
import com.dairyncia.dto.ManagerUpdateDTO;
import com.dairyncia.dto.PendingUserDto;
import com.dairyncia.entities.User;
import com.dairyncia.repository.FarmerRepository;
import com.dairyncia.repository.RoleRepository;
import com.dairyncia.repository.UserRepository;
import com.dairyncia.service.AdminService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/admin")
@PreAuthorize("hasAnyRole('ADMIN','MANAGER')")
@RequiredArgsConstructor
public class AdminController {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final FarmerRepository farmerRepository;
    //private final AddressRepository addressRepository;
    //private final BankDetailsRepository bankDetailsRepository;
    //private final MilkCollectionRepository milkCollectionRepository;
    private final AdminService adminService;

    // ================= ASSIGN ROLE =================
    @PostMapping("/assign-role")
    public ResponseEntity<?> assignRole(@Valid @RequestBody AssignRoleDto dto) {
        ApiResponse response=adminService.assignRole(dto);
        
        return new ResponseEntity<>(response,HttpStatus.OK);
    }
    
    // get all managers
    @GetMapping("/managers")
    public ResponseEntity<List<ManagerResponseDTO>> getManagers(){
    	return ResponseEntity.ok(adminService.getAllManagers());
    }
    
    // update manager
    @PutMapping("/managers/{managerId}")
    public ResponseEntity<ApiResponse> updateManager(
    	@PathVariable Long managerId,
    	@RequestBody ManagerUpdateDTO dto){
    	
    	adminService.updateManager(managerId,dto);
    	
    	
    	return ResponseEntity.ok(new ApiResponse("Manager Updated Successfully","SUCCESS")); 
    }
    
    // delete manager by id
    @DeleteMapping("/managers/{managerId}")
    public ResponseEntity<ApiResponse> deleteManager(@PathVariable Long managerId){
    	adminService.deleteManager(managerId);
    	
    	return ResponseEntity.ok(new ApiResponse("Manager Deleted Successfully","SUCCESS")); 
    }
    
    @GetMapping("/farmers")
    public List<FarmerResponseDTO> getAllFarmers() {
        return adminService.getAllFarmers();
    }
    
   

 // Get Farmer by ID
    @GetMapping("/farmers/{farmerId}")
    public FarmerResponseDTO getFarmer(@PathVariable Long farmerId) {
        return adminService.getFarmerById(farmerId);
    }

    // Update Farmer profile
    @PutMapping("/farmers/{farmerId}")
    public ApiResponse updateFarmer(
            @PathVariable Long farmerId,
            @RequestBody FarmerProfileUpdateDTO dto) {
        return adminService.updateFarmer(farmerId, dto);
    }
    
    // editing the addresses
    @PostMapping("/farmers/{userId}/address")
    public ApiResponse saveOrupdateAddress(
    		@PathVariable Long userId, @RequestBody AddressDTO dto) {
    	return adminService.saveOrUpdateFarmerAddress(userId,dto);
    }
    
    // editing the bank-details
    @PostMapping("/farmers/{userId}/bank")
    public ApiResponse saveOrupdateBankDetails(
    		@PathVariable Long userId, @RequestBody BankDetailsDTO dto) {
    	return adminService.saveOrUpdateBankDetails(userId,dto);
    }
    
    // delete farmer by id
    @DeleteMapping("/farmers/{userId}")
    public ApiResponse deleteFarmer(@PathVariable Long userId) {
        return adminService.deleteFarmerByUserId(userId);
    }
    
    @GetMapping("/pending-users")
    public ResponseEntity<List<PendingUserDto>> getPendingUsers() {
        List<User> pendingUsers = userRepository.findPendingUsers();
        
        List<PendingUserDto> result = new ArrayList<>();
        for (int i = 0; i < pendingUsers.size(); i++) {
            User u = pendingUsers.get(i);
            result.add(new PendingUserDto(i+1,u.getId(),u.getEmail(),u.getFullName(),u.getPhoneNumber()));
        }
        
        return ResponseEntity.ok(result);
    }

    
     



    // ================= PENDING USERS =================
    /*@GetMapping("/pending-users")
    public ResponseEntity<List<PendingUserDto>> getPendingUsers() {
        List<User> pendingUsers = userRepository.findPendingUsers();
        
        List<PendingUserDto> result = new ArrayList<>();
        for (int i = 0; i < pendingUsers.size(); i++) {
            User u = pendingUsers.get(i);
            result.add(new PendingUserDto(i+1,u.getId(),u.getEmail(),u.getFullName(),u.getPhoneNumber()));
        }
        
        return ResponseEntity.ok(result);
    }

    // ================= FARMERS =================
    @GetMapping("/farmers")
    public ResponseEntity<List<FarmerListDTO>> getAllFarmers() {
        List<Farmer> farmers = farmerRepository.findAllWithUser();
        
        List<FarmerListDTO> result=farmers.stream()
        		.map(FarmerListDTO::new)
        		.toList();
        
        return ResponseEntity.ok(result);
    }

    @GetMapping("/farmers/{id}")
    public ResponseEntity<FarmerDetailsDto> getFarmerDetails(@PathVariable Integer id) {
        Farmer farmer = farmerRepository.findByIdWithDetails(id)
            .orElseThrow(() -> new ResourceNotFoundException("Farmer not found"));

        BankDetails bank = bankDetailsRepository.findByUserId(farmer.getUserId())
            .orElse(null);

        FarmerDetailsDto dto = FarmerDetailsDto.builder()
            .farmerId(farmer.getId())
            .userId(farmer.getUserId())
            .fullName(farmer.getUser().getFullName())
            .email(farmer.getUser().getEmail())
            .phone(farmer.getUser().getPhoneNumber())
            .build();

        if (farmer.getAddress() != null) {
            dto.setAddress(AddressDto.fromEntity(farmer.getAddress()));
        }

        if (bank != null) {
            dto.setBankDetails(BankDetailsDto.fromEntity(bank));
        }

        return ResponseEntity.ok(dto);
    }

    @PutMapping("/farmers/{id}")
    public ResponseEntity<?> updateFarmer(
            @PathVariable Integer id,
            @Valid @RequestBody UpdateFarmerDto dto) {
        
        Farmer farmer = farmerRepository.findByIdWithUser(id)
            .orElseThrow(() -> new ResourceNotFoundException("Farmer not found"));

        farmer.getUser().setFullName(dto.getFullName());
        farmer.getUser().setPhoneNumber(dto.getPhoneNumber());
        
        userRepository.save(farmer.getUser());

        return ResponseEntity.ok("Farmer profile updated successfully");
    }

    @DeleteMapping("/farmers/{id}")
    public ResponseEntity<?> deleteFarmer(@PathVariable Integer id) {
        Farmer farmer = farmerRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Farmer not found"));

        // Check if farmer has milk collections
        boolean hasMilk = milkCollectionRepository.existsByFarmerId(id);
        if (hasMilk) {
            throw new BadRequestException("Farmer cannot be deleted. Milk records exist.");
        }

        farmerRepository.delete(farmer);
        return ResponseEntity.ok("Farmer deleted");
    }

    // ================= MANAGERS =================
    @GetMapping("/managers")
    public ResponseEntity<List<ManagerListDto>> getAllManagers() {
        Role managerRole = roleRepository.findByName(Role.RoleType.ROLE_MANAGER)
            .orElseThrow(() -> new ResourceNotFoundException("Manager role not found"));

        List<User> managers = userRepository.findByRolesContaining(managerRole);

        List<ManagerListDto> result = managers.stream()
            .map(u -> ManagerListDto.builder()
                .managerId(u.getId())
                .email(u.getEmail())
                .fullName(u.getFullName())
                .phone(u.getPhoneNumber())
                .build())
            .collect(Collectors.toList());

        return ResponseEntity.ok(result);
    }

    @GetMapping("/managers/{id}")
    public ResponseEntity<?> getManagerById(@PathVariable String id) {
        User manager = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Manager not found"));

        Map<String, Object> response = new HashMap<>();
        response.put("id", manager.getId());
        response.put("email", manager.getEmail());
        response.put("fullName", manager.getFullName());
        response.put("phoneNumber", manager.getPhoneNumber());

        return ResponseEntity.ok(response);
    }

    @PutMapping("/managers/{id}")
    public ResponseEntity<?> updateManager(
            @PathVariable String id,
            @Valid @RequestBody UpdateManagerDto dto) {
        
        User manager = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Manager not found"));

        manager.setFullName(dto.getFullName());
        manager.setPhoneNumber(dto.getPhoneNumber());
        
        userRepository.save(manager);

        return ResponseEntity.ok("Manager updated successfully");
    }

    @DeleteMapping("/managers/{id}")
    public ResponseEntity<?> deleteManager(@PathVariable String id) {
        User manager = userRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Manager not found"));

        // Check if manager has milk collection records
        boolean isUsed = milkCollectionRepository.existsByManagerId(id);
        if (isUsed) {
            throw new BadRequestException("Manager cannot be deleted. Records exist.");
        }

        userRepository.delete(manager);
        return ResponseEntity.ok("Manager deleted successfully");
    }*/

    // Continue in next artifact...
}